src/qt/rpcconsole.cpp

```diff
@@ -688,8 +689,10 @@ void RPCConsole::setClientModel(ClientModel *model)
     if (!model) {
         // Client model is being set to 0, this means shutdown() is about to be called.
         // Make sure we clean up the executor thread
+        QEventLoop loop;
+        connect(&thread, &QThread::finished, &loop, &QEventLoop::quit);
         Q_EMIT stopExecutor();
-        thread.wait();
+        if (!thread.isFinished()) loop.exec();
```

Comment:
You're implicitly using `QueuedConnection` here: https://github.com/bitcoin/bitcoin/blob/9255963890f07f93646cc32703da935dcb24d454/src/qt/rpcconsole.cpp#L693
Therefore,  the `thread.isFinished()` is called immediately after the `stopExecutor()` signal is emitted. If the `thread` is idle the value of `thread.isFinished()` may be unpredictable, IMO.

Reply:
If I understood your concern correctly, you're afraid that the thread.isFinished() result may be unpredictable and as a result, the QEventLoop may never quit since it is tied to the thread finished() signal.

I don't see a problem considering how the QThread lifecycle works. As per documentation, the order is:
* Thread quit() or exit() is called
* As soon as exec completes, isFinished starts to return true
* The finished() signal is emitted

As a result we should consistently get the loop.exec before the thread finished signal is emitted.
Also, if thread.isFinished return true, the code in the if statement is not executed and shutdown continues normally.

Please let me know if I misunderstood your concern.
Thanks,

Reply:
As you noticed, we're using `QueuedConnection` implicitly here.
The EventLoop will only receive the thread finished event after we call `exec()`. I ran multiple tests locally to confirm that. Even when the thread finishes right away and we insert some long running process before the exec, it still exits successfully.
That would in fact be a problem if we were using Qt::DirectConnection.

@promag , do you have anything to add to this comment?  (as you suggested that snippet of code)

Reply:
Agree: forcing event dispatching by means of a local event loop works as expected.